instance/tests/models/test_openedx_appserver.py

```diff
@@ -0,0 +1,391 @@
+    def test_provision(self, mocks):
+        """
+        Run provisioning sequence
+        """
+        mocks.mock_run_ansible_playbooks.return_value = (['log'], 0)
+        mocks.mock_create_server.side_effect = [Mock(id='test-run-provisioning-server'), None]
+        mocks.os_server_manager.add_fixture('test-run-provisioning-server', 'openstack/api_server_2_active.json')
+        mock_reboot = mocks.os_server_manager.get_os_server('test-run-provisioning-server').reboot
+
+        appserver = make_test_appserver()
+        self.assertEqual(appserver.status, AppServerStatus.New)
+        self.assertEqual(appserver.server.status, Server.Status.Pending)
+        result = appserver.provision()
+        self.assertTrue(result)
+        self.assertEqual(appserver.status, AppServerStatus.Running)
+        self.assertEqual(appserver.server.status, Server.Status.Ready)
+        self.assertEqual(mocks.mock_run_ansible_playbooks.call_count, 1)
+        self.assertEqual(mock_reboot.call_count, 1)
+
+    @patch_services
+    def test_provision_build_failed(self, mocks):
+        """
+        Run provisioning sequence failing server creation on purpose to make sure
+        server and instance statuses will be set accordingly.
+        """
+        appserver = make_test_appserver()
+        self.assertEqual(appserver.status, AppServerStatus.New)
+        self.assertEqual(appserver.server.status, Server.Status.Pending)
+
+        mocks.mock_create_server.side_effect = novaclient.exceptions.ClientException(400)
+        result = appserver.provision()
+        self.assertFalse(result)
+
+        self.assertEqual(appserver.status, AppServerStatus.Error)
+        self.assertEqual(appserver.server.status, Server.Status.BuildFailed)
+        mocks.mock_provision_failed_email.assert_called_once_with('Unable to start an OpenStack server')
+
+    @patch_services
+    def test_provision_failed(self, mocks):
+        """
+        Run provisioning sequence failing the deployment on purpose to make sure
+        server and instance statuses will be set accordingly.
+        """
+        log_lines = ['log']
+        mocks.mock_run_ansible_playbooks.return_value = (log_lines, 1)
+        appserver = make_test_appserver()
+        self.assertEqual(appserver.status, AppServerStatus.New)
+        self.assertEqual(appserver.server.status, Server.Status.Pending)
+        result = appserver.provision()
+        self.assertFalse(result)
+        self.assertEqual(appserver.status, AppServerStatus.ConfigurationFailed)
+        self.assertEqual(appserver.server.status, Server.Status.Ready)
+        mocks.mock_provision_failed_email.assert_called_once_with(
+            "AppServer deploy failed: Ansible play exited with non-zero exit code", log_lines
+        )
+
+    @patch_services
+    def test_provision_unhandled_exception(self, mocks):
+        """
+        Make sure that all servers are terminated if there is an unhandled exception during
+        provisioning.
```

Comment:
@bradenmacdonald This docstring seems outdated -- there's no logic for checking whether servers are terminated in the method body.